### PR TITLE
frontend: resourceMap: graphGrouping: Avoid merging via RWX PVC

### LINKED ISSUE #4310: RWX PVC in Map breaks groups
## Describe the bug
When creating an RWX (Read Write Many) and attaching it to multiple pods on differents deployments, will make the all the pods connected to the PVC on the same group. 

## To Reproduce
Steps to reproduce the bug:
1. Create a RWX PVC
2. Attach it to two pods in two differents deployments
3. Head over to the Map
4. See the deployments are moved in a single group (somtime the Deployment stays alone and only the pod is moved, so one deployment have nothing and the other have both pods)

## Environment (please provide info about your environment):
 - Installation type: In-Cluster with Helm
 - Headlamp Version: 0.38.0, 0.39.0 (upgraded after issue creation)

## Are you able to fix this issue?
No. (i'm not a react developer)
> Do you want to help fix this issue? Please see our [contribution docs](https://headlamp.dev/docs/latest/contributing). For any non-trivial contributions, we advise you to join our [our slack channel](https://kubernetes.slack.com/messages/headlamp) to discuss the implementation with Headlamp's core developers.

## Additional Context
Here is how this shows with Traefik. I have two demonset (which where in two groups until i added the PVC and linked all pods to it)

Namespace view :
Image

Sync Certificate DaemonSet view : 
Image

After more thinking i guess it's to avoid duplication of the nodes but for RWX PVC i think it could be duplicated or showing it in another way, because in extreme case, it can be hard to view anything.

Appwrite API Deployment view :
Image

### PR DESCRIPTION
**Pull request by rootp1**
## Summary

Fixes Resource Map grouping when multiple unrelated workloads share a `ReadWriteMany` (RWX) `PersistentVolumeClaim`. Previously, a shared RWX PVC could bridge otherwise unrelated components into a single group. This adds a generic `isNonGrouping` edge flag (per review feedback from @sniok, instead of hardcoding PVC-specific logic in `graphGrouping`), and marks PVC→Pod edges as non-grouping when the PVC is RWX. Nodes only reachable via non-grouping edges to more than one neighbor are split out of connected-component detection and cloned back into each component they touch, so every group keeps showing its own copy of the shared resource without merging with unrelated workloads.

This is a continuation of #4499 (by @devsubid), rebased against latest `main` and reworked to:
- Use the generic `isNonGrouping` edge flag suggested by @sniok instead of PVC-specific grouping logic.
- Address outstanding Copilot review comments (JSDoc on new helpers, avoid unnecessary `as any`, fix edge direction in tests, avoid redundant checks).
- Rebase cleanly onto current `main`, which has since gained a BFS/perf-instrumented `getConnectedComponents` — the fix is adapted to preserve that instrumentation.

Fixes #4310

## Screenshots

Two unrelated Deployments (`web-frontend`, `report-generator`) mounting the same RWX PVC (`shared-uploads`), same live cluster, same namespace, same viewport — only this fix applied differs between the two:

**Before** — both workloads collapse into a single connected group, bridged by the shared PVC:

Before the fix: web-frontend and report-generator merged into one group via the shared RWX PVC

**After** — each workload stays in its own group, each with its own copy of the PVC node:

After the fix: web-frontend and report-generator in two separate groups, each with its own shared-uploads PVC node

### CODE REVIEW

**Comment by rootp1** [comment]
**Before**  the shared RWX PVC bridges both otherwise-unrelated Deployments into a single connected group:



**After** each Deployment stays in its own separate group, each with its own copy of the PVC node:

File: frontend/src/components/resourceMap/graph/graphGrouping.tsx [hunk 1]
```diff
@@ -38,121 +38,291 @@ export const getGraphSize = (graph: GraphNode) => {
...
+  });
+
+  const baseGraphLookup = makeGraphLookup(baseNodes, baseEdges);
+  const baseComponents = getConnectedComponentGroups(baseNodes, baseGraphLookup);
+
+  baseComponents.forEach(component => {
+    const componentNodeIds = new Set(component.nodes?.map(n => n.id) ?? []);
+
+    sharedNodeIds.forEach(sharedId => {
+      const sharedNode = sharedNodesById.get(sharedId);
```
[the remark this answers is not in the document]
Comment (rootp1) [reply frontend/src/components/resourceMap/graph/graphGrouping.tsx]: Fixed in 6660b19: replaced the per-component scan of every shared node with a single-pass `componentByNodeId` map built once over the base components, so each shared node's incident edges resolve to their target component(s) directly instead of every component scanning every shared node.

File: frontend/src/components/resourceMap/graph/graphGrouping.tsx [hunk 2]
```diff
@@ -38,121 +38,291 @@ export const getGraphSize = (graph: GraphNode) => {
...
+        const otherNodeId = edge.source === sharedId ? edge.target : edge.source;
+        return componentNodeIds.has(otherNodeId);
+      });
+
+      if (edgesToThisComponent.length === 0) {
+        return;
+      }
+
+      const clonedId = `${sharedId}--${component.id}`;
+      component.nodes = [...(component.nodes ?? []), { ...sharedNode, id: clonedId }];
```
[the remark this answers is not in the document]
Comment (rootp1) [reply frontend/src/components/resourceMap/graph/graphGrouping.tsx:318]: Fixed in 6660b19: a shared node is now only cloned with a synthetic per-component ID when its incident edges resolve to more than one distinct base component. When they all resolve to a single component (e.g. an RWX PVC mounted by two Pods of the same Deployment), the original node and edge IDs are reattached unchanged, so `?node=<uid>`/`defaultNodeSelection` exact-ID lookups keep working. Added a regression test for this single-component case.

File: frontend/src/components/resourceMap/graph/graphModel.tsx [hunk 3]
```diff
@@ -82,6 +82,13 @@ export interface GraphEdge {
   label?: ReactNode;
   /** Custom data for this node */
   data?: any;
+  /**
+   * If set to true, this edge will not cause its source and target nodes to be
+   * grouped together. Useful for "shared" resources that are referenced by
+   * multiple otherwise-unrelated components (e.g. a ReadWriteMany PVC mounted
+   * by several Deployments) so they don't get merged into a single group.
+   */
```
[the remark this answers is not in the document]
Comment (rootp1) [reply frontend/src/components/resourceMap/graph/graphModel.tsx:98]: Fixed in 6660b19: reworded the `isNonGrouping` doc comment to state the actual rule — it only affects grouping when *every* incident edge on a node is flagged and the node touches more than one distinct neighbor; a node with even a single normal edge is grouped as usual regardless of any other flagged edges.

File: frontend/src/components/resourceMap/sources/GraphSources.tsx [hunk 4]
```diff
@@ -349,6 +349,7 @@ export function GraphSourceManager({ sources, children, relations }: GraphSource
                 source: from.id,
                 target: to.id,
                 label: relation.label,
+                ...relation.edgeAttributes?.(from, to),
```
[the remark this answers is not in the document]
Comment (rootp1) [reply frontend/src/components/resourceMap/sources/GraphSources.tsx]: Added in 6660b19: a new GraphSourceManager test that passes a relation with an `edgeAttributes` callback and asserts the returned `isNonGrouping` flag actually lands on the generated edge, exercising the real relation-to-edge propagation path in GraphSources.tsx rather than constructing flagged edges directly.

**Comment by rootp1** [comment]
@illume ALL the review comments are addressed

**Review by illume — CHANGES_REQUESTED** [review]
Thanks for working on this.


Would you mind addressing the open Copilot review comments? Please mark each comment as resolved after addressing it.

File: frontend/src/components/resourceMap/graph/graphGrouping.tsx [hunk 5]
```diff
@@ -38,121 +38,318 @@ export const getGraphSize = (graph: GraphNode) => {
...
+    if (allEdges.length === 0 || allEdges.some(edge => !edge.isNonGrouping)) {
+      return;
+    }
+
+    const neighborNodeIds = new Set<string>();
+    outgoingEdges.forEach(edge => neighborNodeIds.add(edge.target));
+    incomingEdges.forEach(edge => neighborNodeIds.add(edge.source));
+
+    if (neighborNodeIds.size > 1) {
+      sharedNodeIds.add(node.id);
```
[the remark this answers is not in the document]
Comment (rootp1) [reply frontend/src/components/resourceMap/graph/graphGrouping.tsx:212]: Fixed in 4337dc6 (squashed): replaced the boolean `isNonGrouping` flag with an edge-role-aware `nonGroupingSide?: 'source' | 'target'` on GraphEdge, naming which specific endpoint of an edge is allowed to be treated as shareable (the PVC, not the Pod, for pvc-pod edges). `findSharedNodeIds` now only classifies a node as shared when every incident edge names *that node* as its `nonGroupingSide` — so a standalone Pod mounting two RWX PVCs (edges both naming the PVC side) is correctly never treated as shared, even though it has no 'normal' edges and touches 2 neighbors. Added a regression test for exactly this case ('does not treat a standalone Pod mounting two RWX PVCs as shared') plus a relations.test.tsx case asserting pvc-pod's edgeAttributes returns the flag for RWX and {} for RWO.

**Comment by rootp1** [comment]
@illume all Copilot review comments are now addressed and resolved:

1. **Quadratic scan** — replaced the per-component scan of every shared node with a single-pass node→component map.
2. **Wrong-ID cloning when only one component is touched** — a shared node is now only cloned with a synthetic per-component ID when it actually bridges more than one component; otherwise it's reattached with its original ID (so `?node=<uid>` selection keeps working). Added a regression test.
3. **Overstated `isNonGrouping` doc contract** — went further than rewording it: replaced the boolean with an edge-role-aware `nonGroupingSide?: 'source' | 'target'`, which also fixes the follow-up symmetric-evaluation bug Copilot found in the next review pass (a standalone Pod mounting two RWX PVCs was being misidentified as the shared resource itself). Added a regression test for that case too.
4. **Untested relation→edge propagation** — added a `GraphSourceManager` test asserting a relation's `edgeAttributes` callback actually reaches the generated edge, plus a `relations.test.tsx` case asserting `pvc-pod` returns the flag only for RWX, not RWO.

Also squashed the corrective commits into a single commit (`4337dc6`) per Copilot's suggestion, so the branch presents the final implementation directly.

All 321 resourceMap tests pass, ESLint and `tsc --noEmit` are clean.

**Comment by rootp1** [comment]
Cleaned up the commit history: squashed the long commit body into a single-line subject and added the missing DCO sign-off, no functional changes. Reran the graphGrouping and GraphSources unit tests locally, all 37 passing. @illume please have another look whenever you get a chance.

File: frontend/src/components/resourceMap/sources/GraphSources.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (rootp1) [reply frontend/src/components/resourceMap/sources/GraphSources.tsx]: Fixed in 9922a20 (amended): reordered the edge object so the generated `id`/`source`/`target` are applied after `...relation.edgeAttributes?.(from, to)`, making them authoritative and un-overridable at runtime. Also narrowed `edgeAttributes`'s return type on both `Relation` and `makeRelation` to `Partial<Omit<GraphEdge, 'id' | 'source' | 'target'>>`, so this can't even be attempted at the type level (only `label`, `data`, and `nonGroupingSide` remain customizable). Added a regression test ('does not let a relation edgeAttributes callback override the generated id/source/target') that forces the old runtime hazard via an `as unknown as` cast, since a dynamically-loaded plugin relation could still return the wrong shape despite the type.

**Comment by rootp1** [comment]
@illume no more comments to address, for the final pointout, I reordered the edge object

**Review by illume — APPROVED** [review]
🎉 thanks!
